mess/scf.py

```diff
@@ -10,14 +12,121 @@
 from mess.orthnorm import cholesky
 from mess.types import OrthNormTransform
 
+from typing import Tuple, Literal
+
+# TODO: remove or add centralized typing.py file
+FloatBxB = jnp.ndarray
+FloatSCF = jnp.ndarray
+FloatSCFxBxB = jnp.ndarray
+FloatSCFxSCF = jnp.ndarray
+
+
+def compute_residual(F: FloatBxB, P: FloatBxB, overlap: FloatBxB) -> FloatBxB:
+    """
+    Computes the residual matrix for the Fock matrix.
+    F: Fock matrix
+    P: Density matrix
+    cst: constant system tensors containing the overlap matrix
+    """
+    temp = jnp.einsum("ab,bc,cd->ad", F, P, overlap)
+    X = cholesky(overlap)  # TODO: avoid unnecessary repomputation
+    res = X.T @ (temp - temp.T) @ X
+    res = (res - res.T) / 2  # Recover anti-symmetry violated by numerical errors
+    return res
+
+
+@jax.jit
+def solve_pulay_equation(current_cycle: int, overlap: FloatSCFxSCF) -> FloatSCF:
+    B = overlap
+    total_cycles = overlap.shape[0]
+    constraint_idx = current_cycle + 1
+    set_vec = -1 * (jnp.arange(total_cycles) < constraint_idx)
+    B = B.at[:, constraint_idx].set(set_vec)
+    B = B.at[constraint_idx, :].set(set_vec)
+    B = B.at[constraint_idx, constraint_idx].set(0)
+    B = (B + B.T) / 2  # Ensure symmetry
+    rhs = jnp.zeros(total_cycles).at[constraint_idx].set(-1)
+    fock_coeffs = jax.scipy.linalg.solve(
+        B, rhs, assume_a="sym"
+    )  # (x0, ..., x_{n-1}, lambda, 0, ..., 0)
+    return fock_coeffs
+
+
+@eqx.dataclass  # TODO: not sure wether this works, this used to be a flax dataclass
+class DiisState:
+    overlap: FloatBxB  # do not confuse with the basis set overlap matrix
```

Comment:
not the best naming convention but in the DIIS paper this is called overlap too as far as I remember